src/librustc_mir_build/hair/pattern/const_to_pat.rs

```diff
@@ -280,4 +265,81 @@ impl<'a, 'tcx> ConstToPat<'a, 'tcx> {
 
         Pat { span, ty: cv.ty, kind: Box::new(kind) }
     }
+
+    fn ty_has_partial_eq_impl(&self, ty: Ty<'tcx>) -> bool {
+        let tcx = self.tcx();
+
+        let is_partial_eq = |ty| {
+            let partial_eq_trait_id = tcx.require_lang_item(EqTraitLangItem, Some(self.span));
+            let obligation: PredicateObligation<'_> = predicate_for_trait_def(
+                tcx,
+                self.param_env,
+                ObligationCause::misc(self.span, self.id),
+                partial_eq_trait_id,
+                0,
+                ty,
+                &[],
+            );
+
+            // FIXME: should this call a `predicate_must_hold` variant instead?
+            self.infcx.predicate_may_hold(&obligation)
+        };
+
+        // Higher-ranked function pointers, such as `for<'r> fn(&'r i32)` are allowed in patterns
+        // but do not satisfy `Self: PartialEq` due to shortcomings in the trait solver.
+        // Check for bare function pointers first since it is cheap to do so.
+        if let ty::FnPtr(_) = ty.kind {
+            return true;
+        }
+
+        // In general, types that appear in patterns need to implement `PartialEq`.
+        if is_partial_eq(ty) {
+            return true;
+        }
+
+        // HACK: The check for bare function pointers will miss generic types that are instantiated
+        // with a higher-ranked type (`for<'r> fn(&'r i32)`) as a parameter. To preserve backwards
+        // compatibility in this case, we must continue to allow types such as `Option<fn(&i32)>`.
+        //
+        //
+        // We accomplish this by replacing *all* late-bound lifetimes in the type with concrete
+        // ones. This leverages the fact that function pointers with no late-bound lifetimes do
+        // satisfy `PartialEq`. In other words, we transform `Option<for<'r> fn(&'r i32)>` to
+        // `Option<fn(&'erased i32)>` and again check whether `PartialEq` is satisfied.
+        // Obviously this is too permissive, but it is better than the old behavior, which
```

Comment:
It would probably be a good idea to construct an example of something that runs into this (and presumably causes the same ICE that this is fixing most other occurrences of), and then file a bug about fixing *that* remaining hole, right?

@ecstatic-morse , can you provide a piece of example code that runs into that scenario?

Reply:
My thought was that you could write something like `impl PartialEq for &'static MyStruct` and then somehow write a higher-ranked reference to `MyStruct` in a `const`. However, I think the only way we let the user write higher-ranked types is with `fn(&i32)` or `Fn(&i32)`, neither of which can run into that problem. Additionally, users can no longer use trait objects in patterns since #71038. If these really are all the ways that the user can write higher-ranked types, perhaps this exactly the right amount of permissive?